src/main/cpp/src/string_to_float_cudf.cuh

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/strings/detail/convert/is_float.cuh>
+#include <cudf/strings/string_view.cuh>
+
+#include <cmath>
+#include <limits>
+
+namespace spark_rapids_jni {
+namespace detail {
+
+/**
+ * @brief This function converts the given string into a
+ * floating point double value.
+ *
+ * This will also map strings containing "NaN", "Inf", etc.
+ * to the appropriate float values.
+ *
+ * This function will also handle scientific notation format.
+ *
+ * This function is a copy of cudf::strings::detail::stod with
+ * the namespace changed to spark_rapids_jni::detail and fixed
+ * an overflow bug of `exp_ten`. It is a short-term solution to
+ * resolve a bug in get_json_object. We should remove this file
+ * once the bug is fixed in cudf in long term.
+ * This diff is `if (exp_ten >= 1e8) break;`
+ */
+__device__ inline double stod(cudf::string_view const& d_str)
+{
+  char const* in_ptr = d_str.data();
+  char const* end    = in_ptr + d_str.size_bytes();
+  if (end == in_ptr) return 0.0;
+  double sign{1.0};
+  if (*in_ptr == '-' || *in_ptr == '+') {
+    sign = (*in_ptr == '-' ? -1 : 1);
+    ++in_ptr;
+  }
+
+  constexpr double infinity      = std::numeric_limits<double>::infinity();
+  constexpr uint64_t max_holding = (std::numeric_limits<uint64_t>::max() - 9L) / 10L;
+
+  // special strings: NaN, Inf
+  if ((in_ptr < end) && *in_ptr > '9') {
+    auto const inf_nan = cudf::string_view(in_ptr, static_cast<cudf::size_type>(end - in_ptr));
+    if (cudf::strings::detail::is_nan_str(inf_nan)) { return nan(""); }
+    if (cudf::strings::detail::is_inf_str(inf_nan)) { return sign * infinity; }
+  }
+
+  // Parse and store the mantissa as much as we can,
+  // until we are about to exceed the limit of uint64_t
+  uint64_t digits = 0;
+  int exp_off     = 0;
+  bool decimal    = false;
+  while (in_ptr < end) {
+    char ch = *in_ptr;
+    if (ch == '.') {
+      decimal = true;
+      ++in_ptr;
+      continue;
+    }
+    if (ch < '0' || ch > '9') break;
+    if (digits > max_holding)
+      exp_off += (int)!decimal;
+    else {
+      digits = (digits * 10L) + static_cast<uint64_t>(ch - '0');
+      if (digits > max_holding) {
+        digits = digits / 10L;
+        exp_off += (int)!decimal;
+      } else
+        exp_off -= (int)decimal;
+    }
+    ++in_ptr;
+  }
+  if (digits == 0) { return sign * static_cast<double>(0); }
+
+  // check for exponent char
+  int exp_ten  = 0;
+  int exp_sign = 1;
+  if (in_ptr < end) {
+    char ch = *in_ptr++;
+    if (ch == 'e' || ch == 'E') {
+      if (in_ptr < end) {
+        ch = *in_ptr;
+        if (ch == '-' || ch == '+') {
+          exp_sign = (ch == '-' ? -1 : 1);
+          ++in_ptr;
+        }
+        while (in_ptr < end) {
+          ch = *in_ptr++;
+          if (ch < '0' || ch > '9') break;
+          exp_ten = (exp_ten * 10) + (int)(ch - '0');
+          if (exp_ten >= 1e8) break;
```

Comment:
Can we have an explanation about why this is here? This is the only change between CUDF and this version so I would like to understand it better. Mainly because it has a magic number of 1e8 in it and I would really like to understand where that came from.